text/0000-nested-publish.md

```diff
@@ -0,0 +1,303 @@
+
+Nested packages may contain other dependencies on nested packages, and these too are included in the published package.
+
+When a group of packages is published in this way, this has a number of useful effects for its dependents (which are not things that Cargo explicitly implements, just consequences of the system):
+
+*   The packages are a single unit for all versioning purposes; there is no way for a version mismatch to arise among them since all the code was published together. Version resolution does not apply (in the same way that it does not for any other `path =` dependency).
+
+    This allows package authors to avoid needing to think about SemVer correctness for their nested packages.
+*   The sub-package is effectively “private” (in a sense like the Rust language's visibility system): it cannot be named as a dependency by any other package on `crates.io`, only by its parent package and sibling sub-packages. The parent package may still re-export items from it, or even the entire crate, in the same ways as it could do with a dependency on a normally published package.
+
+## Example: trait and derive macro
+
+Suppose we want to declare a trait-and-derive-macro package. We can do this as follows. The parent package would have this manifest `foo/Cargo.toml`:
+
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2021"
+publish = true
+
+[dependencies]
+foo-macros = { path = "macros", publish = "nested" } # new syntax
+```
+
+The sub-package manifest `foo/macros/Cargo.toml`:
+
+```toml
+[package]
+name = "macros"            # this name need not be claimed on crates.io
+# version = "0.0.0"        # version number is not used and may be omitted
+edition = "2021"
+publish = "nested"         # new syntax
+
+[lib]
+proc-macro = true
+```
+
+Then you can `cargo publish` from within the parent directory `foo/`, and this will create a single `foo` package on `crates.io`, with no `macros` (or `foo-macros`) package visible except when inspecting the source code or in compilation progress messages.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Cargo: Manifest
+
+Two new possible values are added to the manifest.
+
+* The `package.publish` field allows `"nested"` as a value, in addition to existing `false` and `true`. This value affects `cargo publish` and nested dependencies as described below.
+* The `dependencies.*.publish` field is newly defined, with the only currently allowed value being `"nested"`, to declare that that dependency is a nested dependency. (If desired, `publish = false` could be also be permitted to explicitly document an intent not to nest; this would be identical to the status quo and to omitting the key.)
```

Comment:
`true`/`false` could represent dependencies to strip when published.

Reply:
I'm not sure whether you mean "let's make `publish = false` strip the dependency" or "`publish = false` could be misinterpreted to mean stripping the dependency". I agree with the latter now that it's pointed out; the former would be an incomplete feature because making a package still compilable with removed dependencies would, in many cases, also mean removing some feature declarations.

Reply:
Cargo today strips some dependencies on publish (and yes, cleaning up of features was overlooked and there is an open issue).

I think it would be reasonable for us to have an explicit mechanism to talk about the existing implicit behavior.  This could always be a future possibility within the scope of this RFC.

Reply:
OK, so `publish = false` is *plausibly* “when publishing, delete this dependency”. I think I'd like to not specify that as part of this RFC, though, since as discussed above it's not completely straightforward. I'll change the text to say that no values other than `"nested"` are defined by this RFC, and later we can decide to add `publish = false` or `publish = "strip"`.

Reply:
> and later we can decide to add `publish = false` or `publish = "strip"`.

Just as a data point, when I see `strip`, I expect it to have something to do with this construct.

```toml
[profile.whatever]
strip="..."
```

...so I'd throw in for either `publish = false` or something like `publish="omit"`.

Reply:
I would at least like to see it mentioned as a future possibility to show how we can use this field for other cases (to justify why it isn't `nested = true`)

Reply:
> I would at least like to see it mentioned as a future possibility…

Done.